br/pkg/streamhelper/flush_subscriber.go

```diff
@@ -0,0 +1,318 @@
+
+func newSubscription(toStore Store, output chan<- spans.Valued) *subscription {
+	return &subscription{
+		storeID:     toStore.ID,
+		storeBootAt: toStore.BootAt,
+		output:      output,
+	}
+}
+
+func (s *subscription) connect(ctx context.Context, dialer LogBackupService) {
+	err := s.doConnect(ctx, dialer)
+	if err != nil {
+		s.emitError(err)
+	}
+}
+
+func (s *subscription) doConnect(ctx context.Context, dialer LogBackupService) error {
+	log.Info("[log backup subscription manager] Adding subscription.", zap.Uint64("store", s.storeID), zap.Uint64("boot", s.storeBootAt))
+	s.clearError()
+	s.close()
+
+	c, err := dialer.GetLogBackupClient(ctx, s.storeID)
+	if err != nil {
+		return err
+	}
+	cx, cancel := context.WithCancel(ctx)
+	cli, err := c.SubscribeFlushEvent(cx, &logbackup.SubscribeFlushEventRequest{
+		ClientId: uuid.NewString(),
+	})
+	if err != nil {
+		cancel()
+		return err
+	}
+	s.cancel = cancel
+	s.background = spawnJoinable(func() { s.listenOver(cli) })
+	return nil
+}
+
+func (s *subscription) close() {
+	if s.cancel != nil {
+		s.cancel()
+		s.background.WaitTimeOut(1 * time.Minute)
+	}
+	// HACK: don't close the internal channel here,
+	// because it is a ever-sharing channel.
+}
+
+func (s *subscription) listenOver(cli eventStream) {
+	storeID := s.storeID
+	log.Info("[log backup flush subscriber] Listen starting.", zap.Uint64("store", storeID))
+	for {
+		// Shall we use RecvMsg for better performance?
+		// Note that the spans.Full requires the input slice be immutable.
+		msg, err := cli.Recv()
+		if err != nil {
+			log.Info("[log backup flush subscriber] Listen stopped.", zap.Uint64("store", storeID), logutil.ShortError(err))
+			if err == io.EOF || err == context.Canceled {
+				return
+			}
+			s.emitError(errors.Annotatef(err, "while receiving from store id %d", storeID))
+			return
+		}
+
+		for _, m := range msg.Events {
+			s.output <- spans.Valued{
+				Key: spans.Span{
+					StartKey: decodeKey(m.StartKey),
+					EndKey:   decodeKey(m.EndKey),
+				},
+				Value: m.Checkpoint,
+			}
+		}
+		metrics.RegionCheckpointSubscriptionEvent.WithLabelValues(strconv.Itoa(int(storeID))).Add(float64(len(msg.Events)))
+	}
+}
+
+func (f *FlushSubscriber) addSubscription(ctx context.Context, toStore Store) {
+	f.subscriptions[toStore.ID] = newSubscription(toStore, f.eventsTunnel)
+}
+
+func (f *FlushSubscriber) removeSubscription(toStore uint64) {
+	subs, ok := f.subscriptions[toStore]
+	if ok {
+		log.Info("[log backup subscription manager] Removing subscription.", zap.Uint64("store", toStore))
+		subs.close()
+		delete(f.subscriptions, toStore)
+	}
+}
+
+// decodeKey decodes the key from TiKV, because the region range is encoded in TiKV.
+func decodeKey(key []byte) []byte {
+	if len(key) == 0 {
+		return key
+	}
+	// Ignore the timestamp...
+	_, data, err := codec.DecodeBytes(key, nil)
+	if err != nil {
+		log.Warn("the key from TiKV isn't encoded properly", logutil.Key("key", key), logutil.ShortError(err))
+		return key
```

Comment:
the wrong key may confuse the span tree

Reply:
If we return a empty slice, it would be even more confusing🤔

Reply:
Anyway, let's skip the key for safety.